bin/varnishd/cache/cache_vrt_var.c

```diff
@@ -561,12 +561,25 @@ VRT_r_req_##nm(VRT_CTX)							\
 	return (ctx->req->elem);					\
 }
 
+#define REQ_VAR_U(nm, elem, val)					\
+									\
+VCL_VOID								\
+VRT_u_req_##nm(VRT_CTX)							\
+{									\
+	CHECK_OBJ_NOTNULL(ctx, VRT_CTX_MAGIC);				\
+	CHECK_OBJ_NOTNULL(ctx->req, REQ_MAGIC);				\
+	ctx->req->elem = val;						\
+	ctx->req->elem = val;						\
+}
+
 REQ_VAR_R(backend_hint, director_hint, VCL_BACKEND)
 
 REQ_VAR_L(ttl, d_ttl, VCL_DURATION, if (!(arg>0.0)) arg = 0;)
```

Comment:
do we still want to set negative arguments to zero? Could it be a better option to fail?

Reply:
IMO it's reasonable to expect that `req.ttl` can be set to a negative duration, we even do so in a test case:
https://github.com/varnishcache/varnish-cache/blob/e0b164f9f134c0e594df8cb543036a441836da25/bin/varnishtest/tests/v00020.vtc#L150

Failing on a negative duration has a good chance of breaking existing VCL.

Reply:
That is a coverage test.
What are the semantics of a negative duration for `req.ttl` and `req.grace`?

Reply:
The same as `beresp.ttl` and `beresp.grace`:

https://github.com/varnishcache/varnish-cache/blob/e0b164f9f134c0e594df8cb543036a441836da25/bin/varnishd/cache/cache_vrt_var.c#L743-L744

Reply:
I understand that this "has always been like that", but also please remember that for quite some time, varnish had no way of properly failing at runtime (`VRT_fail()`).

The particular lines you are referring to are [from 10 years ago](https://github.com/varnishcache/varnish-cache/commit/71a357ab29bb4ec4f8e671665395e101ad885d2e) and I am still wondering what a negative ttl/grace is supposed to mean. ttl: already expired before it even went into the cache? If those were the semantics, then we would not be allowed to take the respective objects as "just refreshed". grace: I would think a negative value should move an object into grace mode _earlier_ than ttl.

So, as we come across these questions, we should answer them.

Reply:
I totally agree that "has always been like that" is a weak argument, but I personally prefer to limit breakage if I can. It's not obvious to me what a negative TTL/grace should mean, and I think that question is bigger than the scope of this PR.

Reply:
Let's try to cover this quickly during the next bugwash. I am also fine with continuing to ignore negative values, but I'd like to at least ask for opinions.

Reply:
I think the point here is to cap the TTL, and once we run out of TTL the object goes in grace mode if applicable, so once we run out of TTL `req.grace` applies.

This is why in my opinion `req.ttl` should only limit the ability to look up a fresh object, and that negative values that could be the result of some arithmetic expression should be bumped to zero.

Reply:
@dridi maybe I do not understand your answer, but why should we not fail immediately if an attempt is made to set any of the timers to a negative value?

Reply:
If you use an expression instead of a constant to compute the acceptable TTL limit and the result is negative, I think it would be too harsh to fail the transaction.

If your "clever" `req.ttl` formula can yield something negative, then negative should be considered as allowing no TTL, just like zero.